### ISSUE TITLE
Allow optimizers to skip nn.Parameters that have requires_grad=False

### ISSUE THREAD

**Issue by alykhantejani** [pytorch/pytorch#679]
I am trying to implement a gaussian blur as a convolution layer in a network, where the weights do not change. I currently have this:

```Python
class GaussianBlur(nn.Module):
    def __init__(self, kernelSize=5, sigma=1):
        self.weight = nn.Parameter(self._calculate_weights(kernelSize, sigma),
                                   requires_grad=False)

    def forward(self, x):
        return F.conv2d(x, self.weight)
```

This then gives the error `optimizing a parameter that doesn't require gradients` from the optimizer. 

It would be a nice feature to be able to exclude some parameters from the optimizer. 

I also tried changing `self.weight` to a Variable, but then when `.cuda()` is called, the weights are not transferred, resulting in other errors

**Comment by apaszke** [comment]
You could save the weight as a buffer and wrap it in a Variable at every forward. Regarding the optimizers I think it's better to have users explicitly filter out the parameters that don't require grad (trivial using the filter builtin).

**Comment by apaszke** [comment]
So I don't really think that it makes sense to allow such parameters. If you don't want to optimize some tensors, they're not parameters - they're fixed. You probably don't want to count them in. And if you really need to than `optimizer.SGD(filter(lambda p: p.requries_grad, model.parameters()), lr=1e-3)` should do the trick.

**Comment by mdering** [comment]
if you're dumb like me and copy and pasted this with the typo, the correct code is below:
`optimizer.SGD(filter(lambda p: p.requires_grad, model.parameters()), lr=1e-3)`

**Comment by el3ment** [comment]
@apaszke I am having a very hard time finding an example of how to "wrap it as a buffer" -- can you point me to an implementation of this?

**Comment by fmassa** [comment]
@el3ment here is an example

```python
class MyModule(Module):
    def __init__(self, buf):
        super(MyModule, self).__init__()
        self.register_buffer('buf', buf)

    def forward(self, input):
        return input * Variable(self.buf)
```

**Comment by YannDubs** [comment]
@apaszke although I do agree in most contexts, there can be some issues with `.cuda()`. Indeed by not fixing variables as parameters you have to override the `cuda` and `cpu` methods. I can't say that it's complicated, but I think having constant parameters could help people not think about this.

I encountered this in a very unusual type of network but others might too :)

**Comment by apaszke** [comment]
Buffers are updated when you use `.cuda()` and `.cpu()` on a module, so this shouldn't be an issue.

**Comment by asdfzxh8** [comment]
> if you're dumb like me and copy and pasted this with the typo, the correct code is below:
> `optimizer.SGD(filter(lambda p: p.requires_grad, model.parameters()), lr=1e-3)`

no such thing as dumx or not about it, do any can be perfx, no right wrong

**Comment by Jonbean** [comment]
I encountered the scenario where being able to switch on and off requires_grad will be handy. 
Basically, I would like to freeze the word embedding layer at the very beginning of training (since I use pre-trained embedding, and don't want it to be ruined by the gradients at the beginning of training). I would like to turn on the requires_grad later for the embedding layer when loss is low enough. Any suggestions?
